### PR TITLE
Fix NullPointerException in CO2Calculator.getKiloWattsPerCore

### LINKED ISSUE #109953: NullPointerException in the Profiling plugin (xpack.profiling.action.CO2Calculator.getKiloWattsPerCore)
### Elasticsearch Version

8.15 snapshot, 8.14

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

any

### Problem Description

The Kibana flamegraph and TopN functions view in Universal Profiling stay empty, even though data exists.
The reason for this is a Null Pointer Exception in the profiling plugin caused by incomplete host metadata.

Relevant part of the stacktrace:
```
[2024-06-19T18:37:03,104][WARN ][r.suppressed             ] [runTask-0] path: /_profiling/topn/functions, params: {}, status: 500 java.lang.NullPointerException: Cannot invoke "String.hashCode()" because "<local2>" is null
        at org.elasticsearch.xpack.profiling.action.CO2Calculator.getKiloWattsPerCore(CO2Calculator.java:59)
        at org.elasticsearch.xpack.profiling.action.CO2Calculator.getAnnualCO2Tons(CO2Calculator.java:54)
        at org.elasticsearch.xpack.profiling.action.TransportGetStackTracesAction$StackTraceHandler.calculateCO2AndCosts(TransportGetStackTracesAction.java:631)
        at org.elasticsearch.xpack.profiling.action.TransportGetStackTracesAction$StackTraceHandler.mayFinish(TransportGetStackTracesAction.java:645)
        at org.elasticsearch.xpack.profiling.action.TransportGetStackTracesAction$StackTraceHandler.onHostsResponse(TransportGetStackTracesAction.java:601)
```

### Steps to Reproduce

Use Universal Profiling in Kibana with incomplete entries in `profiling-hosts`.

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by narasimhareddyputta94**
### Description

This pull request fixes a NullPointerException in the CO2Calculator class, which occurred in the `getKiloWattsPerCore` method due to attempts to invoke methods on null values. The issue was reported in #109953.

### Changes Made

- Added null checks in the `getKiloWattsPerCore` method to ensure the method does not attempt to call `hashCode()` on null values.
- Added logging to warn when host metadata or other critical fields are null or incomplete.
- Default values are used to prevent the method from failing when metadata is missing.

### Issue Fixed
Fixes #109953

### Testing
- Tested the changes with various inputs to ensure that the CO2 calculation process does not fail due to missing or incomplete metadata.
- Verified that appropriate warnings are logged when metadata is incomplete.

### Changelog
- Added a changelog entry at `docs/changelog/110020.yaml` to document this fix.

### Labels
- Team:obs-knowledge
- >bug

### Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### Reviewer Request
@elastic/obs-knowledge-team could you please review this pull request?

Please review the changes and let me know if any adjustments are needed. Thank you!

### CODE REVIEW

**Comment by narasimhareddyputta94** [comment]
@elastic/obs-knowledge-team Could someone from the team please assign an appropriate assignee to this PR?

File: x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/action/CO2Calculator.java [hunk 1]
```diff
@@ -43,33 +47,79 @@ final class CO2Calculator {
...
+     * @return The annual CO2 emissions in metric tons
+     */
     public double getAnnualCO2Tons(String hostID, long samples) {
+        // Calculate annual core hours based on sampling duration and frequency
         double annualCoreHours = CostCalculator.annualCoreHours(samplingDurationInSeconds, samples, DEFAULT_SAMPLING_FREQUENCY);
 
+        // Retrieve host metadata for the given hostID
         HostMetadata host = hostMetadata.get(hostID);
         if (host == null) {
+            logger.warning("Host metadata is null for hostID: " + hostID);
```
Comment (rockdaboot) [inline x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/action/CO2Calculator.java:64]: Missing host metadata is expected and may be quite common. Logging this is would be too verbose.

**Review by rockdaboot — COMMENTED** [review]
Thank you for your contribution!

The code doesn't fix all null pointer accesses that can happen with malformed host metadata. The logging should not be done here, it's a pretty hot path under circumstances.

See also #110024 (sorry, I didn't expect someone else to work on it)

The comments you introduced are helpful and highly appreciated. I would be happy to approve them in a separate PR!

**Comment by rockdaboot** [comment]
The issue #109953 has been fixed by #110024.
Please open a separate PR for adding the comments.

**Pull request closed without merging**